tests/test_user.py

```diff
@@ -0,0 +1,119 @@
+    user = User.from_json(
+        user_dict={
+            "id": "someuser",
+            "username": "someuser",
+            "password": hash_password(password),
+            "config": {},
+            "is_admin": False,
+            "services": None,
+        },
+        specter=specter,
+    )
+
+    assert user.encrypted_user_secret is None
+    assert user.plaintext_user_secret is None
+
+    # Even though there's no user_secret yet, the flow calls decrypt anyway...
+    user.decrypt_user_secret(password)
+
+    # ...and a new user_secret is created and stored encrypted and plaintext
+    assert user.encrypted_user_secret is not None
+    assert user.plaintext_user_secret is not None
+
+
+def test_generate_user_secret_on_set_password(empty_data_folder):
+    """
+    Should generate a user_secret if one does not yet exist when the User's password
+    is changed via set_password.
+    """
+    specter = Specter(data_folder=empty_data_folder)
+
+    password = "somepassword"
+    user = User.from_json(
+        user_dict={
+            "id": "someuser",
+            "username": "someuser",
+            "password": hash_password(password),
+            "config": {},
+            "is_admin": False,
+            "services": None,
+        },
+        specter=specter,
+    )
+
+    assert user.encrypted_user_secret is None
+    assert user.plaintext_user_secret is None
+
+    new_password = "mynewpassphrase"
+    user.set_password(new_password)
+
+    assert user.encrypted_user_secret is not None
+    assert user.plaintext_user_secret is not None
+
+    # Reset the plaintext user_secret and test decryption
+    user.plaintext_user_secret = None
+    user.decrypt_user_secret(new_password)
+
+    assert user.plaintext_user_secret is not None
+
+
+def test_reencrypt_user_secret_on_set_password(empty_data_folder):
+    """
+    Should re-encrypt the user_secret when the user changes their password.
+    """
+    specter = Specter(data_folder=empty_data_folder)
+
+    password = "somepassword"
+    user = User.from_json(
+        user_dict={
+            "id": "someuser",
+            "username": "someuser",
+            "password": hash_password(password),
+            "config": {},
+            "is_admin": False,
+            "services": None,
+        },
+        specter=specter,
+    )
+
+    # Force generation of a new user_secret
+    user.decrypt_user_secret(password)
+    assert user.encrypted_user_secret is not None
+    assert user.plaintext_user_secret is not None
+
+    first_encrypted_user_secret = user.encrypted_user_secret
+    first_plaintext_user_secret = user.plaintext_user_secret
+
+    new_password = "mynewpassphrase"
+    user.set_password(new_password)
+
+    # The new encrypted_user_secret will be different...
+    assert first_encrypted_user_secret != user.encrypted_user_secret
+
+    # ...but the plaintext_user_secret remains unchanged
+    assert first_plaintext_user_secret == user.plaintext_user_secret
```

Comment:
Here i see that you have the same assumption than me in my comment above. However i don't understand how it works like this.